tempesta_fw/connection.c

```diff
@@ -90,13 +91,22 @@ tfw_connection_release(TfwConnection *conn)
  * for the life of @conn instance. The socket itself may have been
  * closed, but not deleted. ss_send() makes sure that data is sent
  * only on an active socket.
- *
- * @unref_data is true if we won't use @msg any more.
  */
 int
-tfw_connection_send(TfwConnection *conn, TfwMsg *msg, bool unref_data)
+tfw_connection_send(TfwConnection *conn, TfwMsg *msg)
 {
-	return ss_send(conn->sk, &msg->skb_list, unref_data);
+	int flags = 0;
+	const TfwHttpMsg *hm = (TfwHttpMsg *)msg;
+
+	if (TFW_CONN_TYPE(conn) & Conn_Clnt) {
+		if (hm->flags & TFW_HTTP_CONN_CLOSE)
+			flags |= SS_SEND_F_CONN_CLOSE;
+		if (!(hm->flags & TFW_HTTP_DO_NOT_PASS))
+			flags |= SS_SEND_F_PASS_SKB;
+	}
```

Comment:
Conceptual architecture is broken. Connection layer is generic - it lays under HTTP, HTTPS and probably other protocols, so connection.[ch] shouldn't include HTTP stuff, while http.[ch] include connection.h. So firstly the change introduces circular dependency. Secondly it breaks program modules dependency.

I'd say this is HTTP responsibility to set appropriate connection flags. Moreover, `tfw_connection_send()` is called for server transmissions as well as for client, so there is no sense to introduce the IF construction when the caller knows exactly who it sends data to.